contracts/StakeManager.sol

```diff
@@ -0,0 +1,165 @@
+    }
+
+    function unlockStake() external {
+        address registree = registrees[msg.sender];
+        StakeInfo storage info = stakes[registree];
+        require(registree != address(0), "not owner");
+        require(info.withdrawBlock == 0, "already pending");
+        info.withdrawBlock = block.number + info.unstakeDelay;
+        emit StakeUnlocked(registree, msg.sender, info.withdrawBlock);
+    }
+
+    function withdrawStake() external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        require(stakes[registree].withdrawBlock > 0, "Withdrawal is not scheduled");
+        require(stakes[registree].withdrawBlock <= block.number, "Withdrawal is not due");
+        uint256 amount = stakes[registree].stake;
+        delete stakes[registree];
+        delete registrees[msg.sender];
+        msg.sender.transfer(amount);
+        emit StakeWithdrawn(registree, msg.sender, amount);
+    }
+
+    function authorizeHub(address relayHub) external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        authorizedHubs[registree][relayHub].removalBlock = uint(-1);
+        emit HubAuthorized(registree, relayHub);
+    }
+
+    function unauthorizeHub(address relayHub) external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        RelayHubInfo storage hubInfo = authorizedHubs[registree][relayHub];
+        require(hubInfo.removalBlock == uint(-1), "hub not authorized");
+        uint256 removalBlock = block.number + stakes[registree].unstakeDelay;
+        hubInfo.removalBlock = removalBlock;
+        emit HubUnauthorized(registree, relayHub, removalBlock);
+    }
+
+    function isRegistreeStaked(address registree, uint256 minAmount, uint256 minUnstakeDelay)
+    external
+    view
+    returns (bool) {
+        StakeInfo storage info = stakes[registree];
+        bool isAmountSufficient = info.stake > minAmount;
+        bool isDelaySufficient = info.unstakeDelay > minUnstakeDelay;
+        bool isStakeLocked = info.withdrawBlock == 0;
+        bool isHubAuthorized = authorizedHubs[registree][msg.sender].removalBlock == uint(-1);
+        return
+        isAmountSufficient &&
+        isDelaySufficient &&
+        isStakeLocked &&
+        isHubAuthorized;
+    }
+
+    /// Slash the stake of the relay registree. In order to prevent stake kidnapping, burns half of stake on the way.
+    /// @param registree - entry to penalize
+    /// @param beneficiary - address that receives half of the penalty amount
+    /// @param amount - amount to withdraw from stake
+    function penalizeRegistree(address registree, address payable beneficiary, uint256 amount) external {
+        require(authorizedHubs[registree][msg.sender].removalBlock > block.number, "hub not authorized");
```

Comment:
need first `require(  removalBlock!=0, "not authorized" )`
and then  `require ( removalBlock > block.Number, "authorization expired" )`

Reply:
That would only affect a revert string, why bother? In tests I care about the right outcome, in production you don't get the revert reason in tx receipt anyway

Reply:
just like your require "penalty exceeds stake" below: its redundant (since SafeMath.sub does the same), but gives much better error message.
Better errors are always better. tools are getting there to give more info (e.g. tenderly shows exact error for failed txs)

Reply:
SafeMath errors are more obscure. I will add a separate require if you insist